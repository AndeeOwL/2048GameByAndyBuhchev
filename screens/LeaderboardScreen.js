import AsyncStorage from "@react-native-async-storage/async-storage";
import { useNavigation } from "@react-navigation/native";
import { useState } from "react";
import { ActivityIndicator, Alert, StyleSheet, Text, View } from "react-native";
import { Colors } from "../components/common/Colors";
import { Fonts } from "../components/common/Fonts";
import LeaderboardList from "../components/Leaderboard/LeaderboardList";
import MainLogo from "../components/MainLogo";
import LoginScreen from "./LoginScreen";

function LeaderboardScreen() {
  const [loading, setLoading] = useState(false);
  const [users, setUsers] = useState([]);
  const navigation = useNavigation();

  const login = () => {
    navigation.navigate(LoginScreen);
  };

  const getAllKeys = async () => {
    setLoading(true);
    let keys = [];
    let userArr = [];
    try {
      keys = await AsyncStorage.getAllKeys();
      keys.forEach(async (k) => {
        const usr = await AsyncStorage.getItem(k);
        const userObject = JSON.parse(usr);
        userArr.push(userObject);
      });
      setUsers(...[userArr]);
    } catch (e) {
      Alert.alert("Not found entries", "No users are registered yet");
    }
    const all = await AsyncStorage.getAllKeys();
    console.log(all);
    setLoading(false);
  };

  users.sort((u1, u2) =>
    u1.score < u2.score ? 1 : u1.score > u2.score ? -1 : 0
  );

  if (loading) {
    return (
      <View style={styles.rootContainer}>
        <ActivityIndicator size='large' />
      </View>
    );
  } else {
    return (
      <View style={styles.rootContainer}>
        <MainLogo />
        <Text style={styles.title}>Leaderboard</Text>
        <Text style={styles.buttons} onPress={login}>
          Login
        </Text>
        <Text style={styles.buttons} onPress={getAllKeys}>
          Load Leaderboard
        </Text>
        <LeaderboardList users={users} />
      </View>
    );
  }
}

export default LeaderboardScreen;

const styles = StyleSheet.create({
  activityRoot: {
    flex: 1,
    backgroundColor: Colors.color256,
    alignItems: "center",
    justifyContent: "center",
  },
  rootContainer: {
    flex: 1,
    backgroundColor: Colors.color256,
    alignItems: "center",
  },
  title: {
    fontSize: Fonts.extraBig,
    fontWeight: "bold",
    marginBottom: 15,
  },
  buttons: {
    margin: 15,
    padding: 15,
    backgroundColor: Colors.color8,
    fontSize: Fonts.small,
    fontWeight: "bold",
  },
});
