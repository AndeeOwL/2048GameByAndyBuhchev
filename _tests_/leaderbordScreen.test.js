import { render } from "@testing-library/react-native";
import { fireEvent } from "@testing-library/react-native/build";
import { act } from "react-test-renderer";
import LeaderboardScreen from "../screens/LeaderboardScreen";

const mockedNavigate = jest.fn();

jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
  };
});

describe("Tests SignUpScreen component", () => {
  test("renders correctly", () => {
    const tree = render(<LeaderboardScreen />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("it should render right amount of elements", () => {
    const { getAllByText, getAllByPlaceholderText } = render(
      <LeaderboardScreen />
    );
    const logo = getAllByText("2048");
    expect(logo).toHaveLength(1);
    const text = getAllByText("Leaderboard");
    expect(text).toHaveLength(1);
    const login = getAllByText("Login");
    expect(login).toHaveLength(1);
    const loadLeaderboard = getAllByText("Load Leaderboard");
    expect(loadLeaderboard).toHaveLength(1);
  });

  test("load should be pressable", () => {
    const { getByText } = render(<LeaderboardScreen />);
    const load = getByText("Load Leaderboard");
    fireEvent.press(load);
  });

  test("Pressed login should navigate", () => {
    const { getByText } = render(<LeaderboardScreen />);
    const login = getByText("Login");
    fireEvent.press(login);
    expect(mockedNavigate).toHaveBeenCalledTimes(1);
  });
});
