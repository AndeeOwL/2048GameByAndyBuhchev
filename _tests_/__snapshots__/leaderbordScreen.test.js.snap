// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests SignUpScreen component renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fae873",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#f5db31",
        "height": 150,
        "justifyContent": "center",
        "marginBottom": 50,
        "marginTop": 50,
        "width": 150,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 35,
          "fontWeight": "bold",
        }
      }
    >
      2048
    </Text>
  </View>
  <Text
    style={
      Object {
        "fontSize": 35,
        "fontWeight": "bold",
        "marginBottom": 15,
      }
    }
  >
    Leaderboard
  </Text>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <Text
      onPress={[Function]}
      style={
        Object {
          "backgroundColor": "#f5a164",
          "fontSize": 16,
          "fontWeight": "bold",
          "margin": 15,
          "padding": 15,
        }
      }
    >
      Login
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <Text
      onPress={[Function]}
      style={
        Object {
          "backgroundColor": "#f5a164",
          "fontSize": 16,
          "fontWeight": "bold",
          "margin": 15,
          "padding": 15,
        }
      }
    >
      Load Leaderboard
    </Text>
  </View>
  <RCTScrollView
    style={
      Object {
        "marginBottom": 50,
        "marginHorizontal": 15,
      }
    }
  >
    <View />
  </RCTScrollView>
</View>
`;
